exercises/starknet/basics/starknet4.cairo

```diff
@@ -0,0 +1,151 @@
+// starknet4.cairo
+// Liz, a friend of Jill, wants to manage inventory for her store on-chain.
+// This is a bit challenging for Joe and Jill, Liz prepared an outline
+// for how contract should work, can you help Jill and Joe write it?
+// I AM NOT DONE
+// Execute `starklings hint starknet4` or use the `hint` watch subcommand for a hint.
+
+// I AM NOT DONE
+#[contract]
+mod LizInventory {
+    use starknet::ContractAddress;
+    use starknet::get_caller_address;
+
+    struct Storage {
+        contract_owner: ContractAddress,
+        // TODO: add storage inventory, that maps product_id to stock quantity
+    }
+
+    #[constructor]
+    fn constructor(owner: ContractAddress) {
+        contract_owner::write( owner );
+    }
+
+    #[external]
+    fn add_stock() {
+        // TODO: takes product_id and new_stock
+        // adds new_stock to stock in inventory
+        // only owner can call this
+    }
+
+    #[external]
+    fn purchase() {
+        // TODO: takes product_id and quantity
+        // subtracts quantity from stock in inventory
+        // anybody can call this, assert stock > quantity
+    }
+
+    #[view]
+    fn get_stock() {
+        // TODO: return product stock in inventory
+    }
+
+}
+
+#[cfg(test)]
+mod test {
+    use starknet::ContractAddress;
+    use array::ArrayTrait;
+    use array::SpanTrait;
+    use debug::PrintTrait;
+    use traits::TryInto;
+
+    use starknet::Felt252TryIntoContractAddress;
+    use option::OptionTrait;
+    use super::LizInventory;
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_owner() {
+
+        let owner: felt252 = 'Elizabeth';
+        let owner: ContractAddress = owner.try_into().unwrap();
+        LizInventory::constructor(owner);
+
+        // Check that contract owner is set
+        let contract_owner = LizInventory::contract_owner::read();
+        assert(contract_owner == owner, 'Elizabeth should be the owner');
+    }
+
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_stock() {
+        let owner = util_felt_addr( 'Elizabeth' );
+        LizInventory::constructor(owner);
+
+        // Call contract as owner
+        starknet::testing::set_caller_address( owner );
+
+        // Add stock
+        LizInventory::__external::add_stock( util_span_2_items( 'Nano', 10 ) );
```

Comment:
add stock should be 
`fn add_stock(product_id: u8, new_stock: u128)`  or something like this

but not take a span as input, it makes thing complicated for nothing

Reply:
ah I see, it's because you actually call it from the `__external` namespace, which we shouldn't do

Reply:
product id can be stringlike felt... New stock could be any `ux`

Reply:
Discussed in https://github.com/shramee/starklings-cairo1/pull/103#discussion_r1189186509